internal/grpcapi/create_user.go

```diff
@@ -2,15 +2,61 @@ package grpcapi
 
 import (
 	"context"
+	"database/sql"
+	"errors"
+	"time"
 
-	usersPb "github.com/Sugar-pack/users-manager/pkg/generated/users"
 	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/Sugar-pack/users-manager/internal/db"
+	usersPb "github.com/Sugar-pack/users-manager/pkg/generated/users"
+	"github.com/Sugar-pack/users-manager/pkg/logging"
 )
 
 func (us *UsersService) CreateUser(ctx context.Context, newUser *usersPb.NewUser) (*usersPb.CreatedUser, error) {
+	dbConn := us.dbConn
+	logger := logging.FromContext(ctx)
 	userID := uuid.New()
+	createdAt := time.Now().UTC()
 	txID := uuid.New()
 
+	newDBUser := &db.User{
+		ID:        userID,
+		Name:      newUser.GetName(),
+		CreatedAt: createdAt,
+	}
+
+	tx, err := dbConn.BeginTxx(ctx, nil) // start regular transaction
+	if err != nil {
+		logger.WithError(err).Error("prepare tx failed")
+		return nil, status.Error(codes.Internal, "prepare tx failed")
+	}
+	defer func() {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			if !errors.Is(rollbackErr, sql.ErrTxDone) {
+				logger.WithError(err).Error("rollback tx failed")
+			}
+		}
+	}()
```

Comment:
As i understand defer will use old state(in the state when the defer was called) oy tx variable. Is it safe to use it?

Reply:
defer as a closure will use the latest evaluated value.

defer will evaluate immediately only passed args